game-core/src/main/java/games/strategy/triplea/delegate/UnitComparator.java

```diff
@@ -109,8 +109,10 @@
    */
   public static Comparator<Unit> getUnloadableUnitsComparator(final List<Unit> units, final Route route,
       final PlayerID player) {
-    return Comparator.comparing(TripleAUnit::get,
-        Comparator.nullsLast(getUnloadableTransportsComparator(units, route, player, false)))
+    return Comparator
+        .<Unit, TripleAUnit>comparing(
+            unit -> TripleAUnit.get(unit).getTransportedBy(),
```

Comment:
Not sure why the compiler can't infer the type arguments here.  Even when I simply switched the mapper from the method reference `TripleAUnit::get` to the lambda `unit -> TripleAUnit.get(unit)`, it couldn't infer the types.  Strange as those should be equivalent...

Reply:
Not sure why.

Reply:
same problem with builders, needs the type hinting to know what the first lambda is going to be.  There could be a chain of more operations, like "A -> B -> C". The compiler does know that "C" is the output, but becuase there are potentially mutiple ways to chain to get to a C, it needs to know the starting point.

Reply:
Yeah, I would have thought there being only a single overload for `TripleAUnit#get()`, the compiler would have been able to deduce that the type of the lambda input parameter was `Unit`, but I guess not.  Maybe because `Unit` isn't `final`, it doesn't know if a subtype could be substituted instead?  Dunno...

Anyway, after looking at this again, I realize now I could have just explicitly typed the lambda input parameter, instead of both the input and output type parameters on the `Comparator#comparing()` call:

```java
return Comparator
    .comparing(
        (Unit unit) -> TripleAUnit.get(unit).getTransportedBy(),
        Comparator.nullsLast(getUnloadableTransportsComparator(units, route, player, false)))
    .thenComparing(getMovableUnitsComparator(units, route));
```

Live and learn... :smile: